### PR TITLE
Add CMake options to control compilation flags for AMD GPUs

### PR DESCRIPTION
**Pull request by Rombur**
This PR allows advanced users to set themselves the compiler flags for AMD GPUs. This allows them to set flags that we don't currently expose through CMake.

This PR adds two new CMake options: `Kokkos_IMPL_AMDGPU_FLAGS` and `Kokkos_IMPL_AMDGPU_LINK`. When a user sets `Kokkos_IMPL_AMDGPU_FLAGS`, we do not set any GPU related flags ourselves. It's the responsibility of the user to set flags that make sense. When users set `Kokkos_IMPL_AMDGPU_FLAGS`, they need to explicitly set `Kokkos_ARCH_AMD_GFXYYY`. I feel that if we allowed autodetection, we will end up with hard to find bug. At the moment, we don't check  that the architecture set by the user matches the architecture flag that they set through `Kokkos_IMPL_AMDGPU_FLAGS`.

Related to https://github.com/kokkos/kokkos/issues/7001

cc: @rbberger

### CODE REVIEW

**Comment by Rombur** [comment]
Retest this please

**Comment by Rombur** [comment]
Retest this please

File: cmake/kokkos_arch.cmake [hunk 1]
```diff
@@ -724,6 +729,15 @@ FOREACH(ARCH IN LISTS SUPPORTED_AMD_ARCHS)
   CHECK_AMDGPU_ARCH(${ARCH} ${FLAG})
 ENDFOREACH()
 
+IF(KOKKOS_IMPL_AMDGPU_FLAGS)
+  IF (NOT AMDGPU_ARCH_ALREADY_SPECIFIED)
+    MESSAGE(FATAL_ERROR "When IMPL_AMDGPU_FLAGS is set the architecture autodectection is disabled. "
+      "Please explicitly set the GPU architecture.")
+  ENDIF()
+  GLOBAL_APPEND(KOKKOS_AMDGPU_OPTIONS "${KOKKOS_IMPL_AMDGPU_FLAGS}")
+  GLOBAL_APPEND(KOKKOS_LINK_OPTIONS "${KOKKOS_IMPL_AMDGPU_LINK}")
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:738]: Is that not going to set it as target property and export it?

File: cmake/kokkos_arch.cmake [hunk 2]
```diff
@@ -566,7 +569,7 @@ ENDIF()
 
 #Right now we cannot get the compiler ID when cross-compiling, so just check
 #that HIP is enabled
-IF (KOKKOS_ENABLE_HIP)
+IF (KOKKOS_ENABLE_HIP AND NOT KOKKOS_IMPL_AMDGPU_FLAGS)
```
[the remark this answers is not in the document]
Comment (Rombur) [reply cmake/kokkos_arch.cmake:572]: > Is that our saying we will not deal with any HIP flags if you opt-in?

Yes, that's what I am thinking. If you opt-in, you get a clean state for you to add your flags.

File: cmake/kokkos_arch.cmake [hunk 3]
```diff
@@ -704,13 +707,15 @@ FUNCTION(CHECK_AMDGPU_ARCH ARCH FLAG)
...
       UNSET(KOKKOS_ARCH_${ARCH} PARENT_SCOPE)
     ELSE()
-      IF(KOKKOS_ENABLE_HIP)
-        SET(KOKKOS_HIP_ARCHITECTURES ${FLAG} PARENT_SCOPE)
-      ENDIF()
-      SET(KOKKOS_AMDGPU_ARCH_FLAG ${FLAG} PARENT_SCOPE)
-      GLOBAL_APPEND(KOKKOS_AMDGPU_OPTIONS "${AMDGPU_ARCH_FLAG}=${FLAG}")
-      IF(KOKKOS_ENABLE_HIP_RELOCATABLE_DEVICE_CODE)
-        GLOBAL_APPEND(KOKKOS_LINK_OPTIONS "${AMDGPU_ARCH_FLAG}=${FLAG}")
+      IF(NOT KOKKOS_IMPL_AMDGPU_FLAGS)
```
[the remark this answers is not in the document]
Comment (Rombur) [reply cmake/kokkos_arch.cmake:710]: Ah I forgot about that. I checked that we export the architecture macros correctly but I forgot about this one.

File: cmake/kokkos_arch.cmake (hunk 1 above)
Comment (Rombur) [reply cmake/kokkos_arch.cmake:738]: Yes, that's what I wanted to do.

File: cmake/kokkos_arch.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/kokkos_arch.cmake:572]: Why is the rdc flag different?

File: cmake/kokkos_arch.cmake [hunk 4]
```diff
@@ -571,7 +576,7 @@ IF (KOKKOS_ENABLE_HIP)
     COMPILER_SPECIFIC_FLAGS(
       DEFAULT -fgpu-rdc
     )
-    IF (NOT KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC)
+    IF (NOT KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC AND NOT KOKKOS_IMPL_AMDGPU_FLAGS)
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:579]: I thought we had agreed not to guard anything in that block?

Comment (Rombur) [reply cmake/kokkos_arch.cmake:579]: That's not how I understood it

Comment (dalg24) [reply cmake/kokkos_arch.cmake:579]: I would prefer if you dropped that guard.  Looks good other than that.

**Comment by dalg24** [comment]
@rbberger plz review

**Review by rbberger — APPROVED** [review]
LGTM

**Review by dalg24 — APPROVED** [review]

**Review by nmm0 — APPROVED** [review]
Looks good to me

**Pull request merged**